File: mcp-servers/jvm/src/main/kotlin/io/agentutils/mcp/McpServer.kt
/**
 * agent-mcp-jvm — Java/JVM MCP Server
 *
 * Exposes JVM-specific tooling over JSON-RPC 2.0 via stdio.
 * Tools: jvm_gradle_diagnose, jvm_stacktrace_unpack, jvm_bytecode_signature
 */

package io.agentutils.mcp

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import java.io.BufferedReader
import java.io.InputStreamReader

data class McpRequest(
    val jsonrpc: String = "2.0",
    val id: Any? = null,
    val method: String = "",
    val params: Map<String, Any?>? = null
)

data class McpResponse(
    val jsonrpc: String = "2.0",
    val id: Any? = null,
    val result: Any? = null,
    val error: McpError? = null
)

data class McpError(val code: Int, val message: String)

typealias ToolHandler = (Map<String, Any?>) -> Any

val toolRegistry: Map<String, ToolHandler> = mapOf(
    "jvm_gradle_diagnose"     to ::gradleDiagnose,
    "jvm_stacktrace_unpack"   to ::stacktraceUnpack,
    "jvm_bytecode_signature"  to ::bytecodeSignature
)

val gson = Gson()

fun respond(id: Any?, result: Any? = null, error: McpError? = null) {
    val response = McpResponse(id = id, result = result, error = error)
    println(gson.toJson(response))
    System.out.flush()
}

fun main() {
    System.err.println("[agent-mcp-jvm] started (transport: stdio)")
    val reader = BufferedReader(InputStreamReader(System.`in`))
    val mapType = object : TypeToken<Map<String, Any?>>() {}.type

    for (line in reader.lines()) {
        val trimmed = line.trim()
        if (trimmed.isEmpty()) continue

        val req: McpRequest = try {
            val raw: Map<String, Any?> = gson.fromJson(trimmed, mapType)
            McpRequest(
                id = raw["id"],
                method = raw["method"] as? String ?: "",
                params = raw["params"] as? Map<String, Any?>
            )
        } catch (e: Exception) {
            respond(null, error = McpError(-32700, "Parse error: ${e.message}"))
            continue
        }

        val handler = toolRegistry[req.method]
        if (handler == null) {
            respond(req.id, error = McpError(-32601, "Method not found: ${req.method}"))
            continue
        }

        try {
            val result = handler(req.params ?: emptyMap())
            respond(req.id, result = result)
        } catch (e: Exception) {
            respond(req.id, error = McpError(-32000, e.message ?: "Unknown error"))
        }
    }
}

File: mcp-servers/dotnet/src/McpServer.cs
// agent-mcp-dotnet — .NET / C# MCP Server
//
// Exposes .NET-specific tooling over JSON-RPC 2.0 via stdio.
// Tools: dotnet_roslyn_analyze, dotnet_test_runner, dotnet_nuget_audit

using System.Text.Json;
using System.Text.Json.Serialization;

namespace AgentUtils.MCP;

record McpRequest(
    [property: JsonPropertyName("jsonrpc")] string Jsonrpc,
    [property: JsonPropertyName("id")] JsonElement? Id,
    [property: JsonPropertyName("method")] string Method,
    [property: JsonPropertyName("params")] Dictionary<string, JsonElement>? Params
);

record McpResponse(
    [property: JsonPropertyName("jsonrpc")] string Jsonrpc,
    [property: JsonPropertyName("id")] JsonElement? Id,
    [property: JsonPropertyName("result")] object? Result = null,
    [property: JsonPropertyName("error")] McpError? Error = null
);

record McpError(
    [property: JsonPropertyName("code")] int Code,
    [property: JsonPropertyName("message")] string Message
);

class McpServer
{
    static readonly Dictionary<string, Func<Dictionary<string, JsonElement>?, Task<object?>>> ToolRegistry =
        new()
        {
            ["dotnet_roslyn_analyze"] = RoslynTools.Analyze,
            ["dotnet_test_runner"]    = TestRunnerTools.Run,
            ["dotnet_nuget_audit"]    = NuGetTools.Audit,
        };

    static readonly JsonSerializerOptions JsonOpts = new()
    {
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
        WriteIndented = false,
    };

    static void Respond(JsonElement? id, object? result = null, McpError? error = null)
    {
        var response = new McpResponse("2.0", id, result, error);
        Console.WriteLine(JsonSerializer.Serialize(response, JsonOpts));
    }

    static async Task Main()
    {
        Console.Error.WriteLine("[agent-mcp-dotnet] started (transport: stdio)");

        string? line;
        while ((line = Console.ReadLine()) != null)
        {
            var trimmed = line.Trim();
            if (string.IsNullOrEmpty(trimmed)) continue;

            McpRequest req;
            try
            {
                req = JsonSerializer.Deserialize<McpRequest>(trimmed, JsonOpts)
                    ?? throw new JsonException("Null deserialization result");
            }
            catch (Exception e)
            {
                Respond(null, error: new McpError(-32700, $"Parse error: {e.Message}"));
                continue;
            }

            if (!ToolRegistry.TryGetValue(req.Method, out var handler))
            {
                Respond(req.Id, error: new McpError(-32601, $"Method not found: {req.Method}"));
                continue;
            }

            try
            {
                var result = await handler(req.Params);
                Respond(req.Id, result: result);
            }
            catch (Exception e)
            {
                Respond(req.Id, error: new McpError(-32000, e.Message));
            }
        }
    }
}
